src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,425 @@
+                        $"{originalNumUserWaits} - {originalNumPendingRemoves} == {_numUserWaits}");
+                }
+                finally
+                {
+                    _registeredHandlesLock.Release();
+                    _removesLock.Release();
+                }
+            }
+            
+            /// <summary>
+            /// Queue a call to <see cref="CompleteWait(object)"/> on the ThreadPool.
+            /// </summary>
+            /// <param name="registeredHandle">The handle that completed.</param>
+            /// <param name="timedOut">Whether or not the wait timed out.</param>
+            private void QueueWaitCompletion(RegisteredWait registeredHandle, bool timedOut)
+            {
+                ThreadPool.QueueUserWorkItem(CompleteWait, new CompletedWaitHandle(registeredHandle, timedOut));
+            }
+
+            /// <summary>
+            /// Process the completion of a user-registered wait.
+            /// </summary>
+            /// <param name="state">A <see cref="CompletedWaitHandle"/> object representing the wait completion.</param>
+            private void CompleteWait(object state)
+            {
+                CompletedWaitHandle handle = (CompletedWaitHandle)state;
+                handle.CompletedHandle.CanUnregister.Reset();
+                handle.CompletedHandle.PerformCallback(handle.TimedOut);
+                handle.CompletedHandle.CanUnregister.Set();
+                if (!handle.CompletedHandle.Repeating)
+                {
+                    QueueOrExecuteUnregisterWait(handle.CompletedHandle);
+                }
+            }
+
+            /// <summary>
+            /// Register a wait handle on this <see cref="WaitThread"/>.
+            /// </summary>
+            /// <param name="handle">The handle to register.</param>
+            /// <returns>If the handle was successfully registered on this wait thread.</returns>
+            public bool RegisterWaitHandle(RegisteredWait handle)
+            {
+                _registeredHandlesLock.Acquire();
+                try
+                {
+                    if (_numUserWaits == WaitHandle.MaxWaitHandles - 1)
+                    {
+                        return false;
+                    }
+
+                    _registeredWaits[_numUserWaits] = handle;
+                    _waitHandles[_numUserWaits + 1] = handle.Handle;
+                    _numUserWaits++;
+
+                    handle.WaitThread = this;
+                }
+                finally
+                {
+                    _registeredHandlesLock.Release();
+                }
+
+                _changeHandlesEvent.Set();
+                return true;
+            }
+
+            /// <summary>
+            /// Starts the wait thread.
+            /// </summary>
+            public void Start()
+            {
+                _waitHandles[0] = _changeHandlesEvent;
+                RuntimeThread waitThread = RuntimeThread.Create(WaitThreadStart);
```

Comment:
This could also be in the constructor since it can OOM. In case of any exception, the wait thread list should not be modified before the exception is propagated up.